core/state/pruner/pruner.go

```diff
@@ -0,0 +1,511 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package pruner
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/state/snapshot"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+const (
+	// stateBloomFilePrefix is the filename prefix of state bloom filter.
+	stateBloomFilePrefix = "statebloom"
+
+	// stateBloomFilePrefix is the filename suffix of state bloom filter.
+	stateBloomFileSuffix = "bf.gz"
+
+	// bloomFilterEntries is the estimated value of the number of trie nodes
+	// and codes contained in the state. It's designed for mainnet but also
+	// suitable for other small testnets.
+	bloomFilterEntries = 600 * 1024 * 1024
+
+	// bloomFalsePositiveRate is the acceptable probability of bloom filter
+	// false-positive. It's around 0.01%.
+	//
+	// Check the https://hur.st/bloomfilter/?n=600000000&p=0.0005&m=&k= for
+	// more calculation details.
+	bloomFalsePositiveRate = 0.0005
+
+	// rangeCompactionThreshold is the minimal deleted entry number for
+	// triggering range compaction. It's a quite arbitrary number but just
+	// to avoid triggering range compaction because of small deletion.
+	rangeCompactionThreshold = 100000
+)
+
+var (
+	// emptyRoot is the known root hash of an empty trie.
+	emptyRoot = common.HexToHash("56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")
+
+	// emptyCode is the known hash of the empty EVM bytecode.
+	emptyCode = crypto.Keccak256(nil)
+)
+
+type Pruner struct {
+	db            ethdb.Database
+	stateBloom    *StateBloom
+	homeDir       string
+	trieCacheName string
+	headHeader    *types.Header
+	snaptree      *snapshot.Tree
+}
+
+// NewPruner creates the pruner instance.
+func NewPruner(db ethdb.Database, headHeader *types.Header, homeDir, trieCacheName string, bloomSize uint64) (*Pruner, error) {
```

Comment:
Nit, is this the trie cache `name` or `path`? If it's already resolved, I think it would be more prudent to call it `trieCachePath` to highlight so (both the param name as well as the struct field)

Reply:
it's `name`. but i will change it to path to keep the consistency